### PR TITLE
Extending the spec with one shot action selector programming

### PR DESCRIPTION
**Pull request by konne-google**

### CODE REVIEW

File: proto/p4/v1/p4runtime.proto [hunk 1]
```diff
@@ -264,6 +265,16 @@ message Action {
   repeated Param params = 4;
 }
 
+message ActionProfileActions {
+  repeated ActionProfileAction action_profile_action = 1;
```
Comment (teverman) [inline proto/p4/v1/p4runtime.proto]: Per protobuf style guide, repeated field names should be plural.

File: docs/v1/P4Runtime-Spec.mdk [hunk 2]
```diff
@@ -2320,6 +2325,116 @@ following semantics.
...
+  action {  /* set nexthop 3 */ }
+}
+~ End Prototext
+
+All the tables associated with an action selector may either be
+programmed exclusively with one shots, or exclusively with
+`ActionProfileMember` and `ActionProfileGroup` messages. Programming
+some entries with one shots, and other entries with
+`ActionProfileMember` and `ActionProfileGroup` messages is not
+allowed.
```
Comment (teverman) [inline docs/v1/P4Runtime-Spec.mdk]: What error message does server produce in this case?

Comment (teverman) [inline docs/v1/P4Runtime-Spec.mdk]: Is a server free to support only the one-shot method?

File: docs/v1/P4Runtime-Spec.mdk [hunk 3]
```diff
@@ -2320,6 +2325,116 @@ following semantics.
 - `DELETE`: Delete the group entry and deallocate the `group_id`. The group
   should not be referenced in the table action of any table entry.
 
+### One Shot Action Selector Programming { #sec-oneshot }
+
+P4Runtime supports syntactic sugar to program a table, which is
+implemented with an action selector, in one shot. This means that a
+table entry, an action profile group, and a set of action profile
+members can programmed with a single update message. Using one shots
```
Comment (antoninbas) [inline docs/v1/P4Runtime-Spec.mdk]: typo: "can be programmed"

File: docs/v1/P4Runtime-Spec.mdk [hunk 4]
```diff
@@ -2320,6 +2325,116 @@ following semantics.
...
+has the advanatage that the controller does not need to keep track of
```
Comment (antoninbas) [inline docs/v1/P4Runtime-Spec.mdk]: typo: "s/advanatage/advantage"

File: proto/p4/v1/p4runtime.proto (hunk 1 above)
Comment (konne-google) [reply proto/p4/v1/p4runtime.proto]: Done.

File: docs/v1/P4Runtime-Spec.mdk [hunk 5]
```diff
@@ -2320,6 +2325,116 @@ following semantics.
...
+
+* `action` is one of the actions specified by the table that is being
+  programmed.
+
+* `weight` specifying the probability of the action's selection at
+  runtime.
+
+* `watch` is the controller defined 32-bit port number that the
+  action's liveness depends on. At runtime, the action must be
+  excluded from selection if the this watch port is down.
```
Comment (antoninbas) [inline docs/v1/P4Runtime-Spec.mdk]: typo, extra "the"

File: docs/v1/P4Runtime-Spec.mdk [hunk 6]
```diff
@@ -2320,6 +2325,116 @@ following semantics.
...
+
+Semantically, one shots are equivalent to programming the table entry,
+group, and members individually. The necessary group id and member ids
```
Comment (antoninbas) [inline docs/v1/P4Runtime-Spec.mdk]: I wouldn't say necessary, it's just one possible / likely way of implementing it

File: docs/v1/P4Runtime-Spec.mdk [hunk 7]
```diff
@@ -2320,6 +2325,116 @@ following semantics.
...
+are bound to unused ids.
+
+For example, consider again the action selector defined in the
+beginning of this section. This table could be programmed with the
```
Comment (antoninbas) [inline docs/v1/P4Runtime-Spec.mdk]: Maybe add a reference to the code block instead to be clearer (see http://madoko.org/reference.html#sec-ids-labels)

File: docs/v1/P4Runtime-Spec.mdk [hunk 8]
```diff
@@ -2320,6 +2325,116 @@ following semantics.
...
+following oneshot update:
```
Comment (antoninbas) [inline docs/v1/P4Runtime-Spec.mdk]: "s/oneshot/one shot"

File: docs/v1/P4Runtime-Spec.mdk (hunk 2 above)
Comment (antoninbas) [reply docs/v1/P4Runtime-Spec.mdk]: I would say yes, the spec should say that
 * a server MUST support the one shot method
 * it is recommended for a server to support both methods
I think the "MUST" for the first bullet point is justified by the fact that if you support the id-based method there is no reason not to be able to support the one shot method.

File: docs/v1/P4Runtime-Spec.mdk (hunk 3 above)
Comment (konne-google) [reply docs/v1/P4Runtime-Spec.mdk]: Done.

File: docs/v1/P4Runtime-Spec.mdk (hunk 4 above)
Comment (konne-google) [reply docs/v1/P4Runtime-Spec.mdk]: Done.

File: docs/v1/P4Runtime-Spec.mdk (hunk 5 above)
Comment (konne-google) [reply docs/v1/P4Runtime-Spec.mdk]: Done.

File: docs/v1/P4Runtime-Spec.mdk (hunk 2 above)
Comment (konne-google) [reply docs/v1/P4Runtime-Spec.mdk]: Done.

Comment (konne-google) [reply docs/v1/P4Runtime-Spec.mdk]: Done.

File: docs/v1/P4Runtime-Spec.mdk (hunk 6 above)
Comment (konne-google) [reply docs/v1/P4Runtime-Spec.mdk]: I meant for the "The necessary ids ..." to be part of the semantics. I've changed the text to clarify this.

File: docs/v1/P4Runtime-Spec.mdk (hunk 8 above)
Comment (konne-google) [reply docs/v1/P4Runtime-Spec.mdk]: Done.

**Pull request merged**